pkg/ipcache/metadata.go

```diff
@@ -0,0 +1,406 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2018-2019 Authors of Cilium
+
+package ipcache
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"strings"
+	"sync"
+
+	"github.com/cilium/cilium/pkg/controller"
+	"github.com/cilium/cilium/pkg/identity"
+	"github.com/cilium/cilium/pkg/identity/cache"
+	"github.com/cilium/cilium/pkg/labels"
+	cidrlabels "github.com/cilium/cilium/pkg/labels/cidr"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/source"
+
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	// idMDMU protects the IdentityMetadata map.
+	//
+	// If this mutex will be held at the same time as the IPCache mutex,
+	// this mutex must be taken first and then take the IPCache mutex.
+	idMDMU lock.RWMutex
+	// IdentityMetadata maps IP prefixes (x.x.x.x/32) to their labels.
+	IdentityMetadata = make(map[string]labels.Labels)
+
+	// ErrLocalIdentityAllocatorUninitialized is an error that's returned when
+	// the local identity allocator is uninitialized.
+	ErrLocalIdentityAllocatorUninitialized = errors.New("local identity allocator uninitialized")
+)
+
+// UpsertMetadata upserts a given IP and its corresponding labels associated
+// with it into the IdentityMetadata map. The given labels are not modified nor
+// is its reference saved, as they're copied when inserting into the map.
+func UpsertMetadata(prefix string, lbls labels.Labels) {
+	l := labels.NewLabelsFromModel(nil)
+	l.MergeLabels(lbls)
+
+	idMDMU.Lock()
+	if cur, ok := IdentityMetadata[prefix]; !ok {
+		IdentityMetadata[prefix] = l
+	} else {
+		l.MergeLabels(cur)
+		IdentityMetadata[prefix] = l
+	}
+	idMDMU.Unlock()
+}
+
+// GetIDMetadataByIP returns the associated labels with an IP. The caller must
+// not modifying the returned object as it's a live reference to the underlying
+// map.
+func GetIDMetadataByIP(prefix string) labels.Labels {
+	idMDMU.RLock()
+	defer idMDMU.RUnlock()
+	return IdentityMetadata[prefix]
+}
+
+// InjectLabels injects labels into the IdentityMetadata (IDMD) map. The given
+// source is the source of the caller, as inserting into the IPCache requires
+// where this updated information is coming from.
+
+// Note that as this function iterates through the IDMD, if it detects a change
+// in labels for a given prefix, then this might allocate a new identity. If a
+// prefix was previously associated with an identity, it will get deallocated,
+// so a balance is kept.
+func InjectLabels(src source.Source, updater identityUpdater, triggerer policyTriggerer) error {
+	if IdentityAllocator == nil || !IdentityAllocator.IsLocalIdentityAllocatorInitialized() {
+		return ErrLocalIdentityAllocatorUninitialized
+	}
+
+	if IPIdentityCache.k8sSyncedChecker == nil ||
```

Comment:
`IPIdentityCache.k8sSyncedChecker == nil` is never true

Reply:
Actually in unit tests it can be true.